test/dynamo/test_hooks.py

```diff
@@ -0,0 +1,341 @@
+
+        x1 = torch.ones(4, requires_grad=True)
+        mod_compiled = aot_module_simplified(mod, (x1,), nop)
+        aot_out = mod_compiled(x1)
+        aot_out[0].backward(torch.ones(4))
+
+        x2 = torch.ones(4, requires_grad=True)
+        dynamo_out = torch._dynamo.optimize("aot_eager", nopython=True)(mod)(x2)
+        dynamo_out[0].backward(torch.ones(4))
+
+        self.assertEqual(dynamo_out, aot_out)
+        self.assertEqual(dynamo_out, eager_out)
+
+        self.assertEqual(x0.grad, x1.grad)
+        self.assertEqual(x0.grad, x2.grad)
+
+    def test_intermediary_hooks_same_on_inductor(self):
+        def my_hook(grad, k=0):
+            return grad + k
+
+        class MyMod(torch.nn.Module):
+            def forward(self, x):
+                y = x.mul(2)
+                hook1 = functools.partial(my_hook, k=3)
+                hook2 = functools.partial(my_hook, k=4)
+                y.register_hook(hook1)
+                y.register_hook(hook2)
+                z = y.mul(3)
+                return (z,)
+
+        mod = MyMod()
+        x0 = torch.ones(4, requires_grad=True)
+        eager_out = mod(x0)
+        eager_out[0].backward(torch.ones(4))
+
+        x1 = torch.ones(4, requires_grad=True)
+        mod_compiled = aot_module_simplified(mod, (x1,), nop)
+        aot_out = mod_compiled(x1)
+        aot_out[0].backward(torch.ones(4))
+
+        x2 = torch.ones(4, requires_grad=True)
+        dynamo_out = torch._dynamo.optimize("inductor", nopython=True)(mod)(x2)
+        dynamo_out[0].backward(torch.ones(4))
+
+        self.assertEqual(dynamo_out, aot_out)
+        self.assertEqual(dynamo_out, eager_out)
+
+        self.assertEqual(x0.grad, x1.grad)
+        self.assertEqual(x0.grad, x2.grad)
+
+    def test_no_recompile_on_hook_identity_change(self):
+        def my_hook(grad, k=0):
+            return grad + k
+
+        def my_hook2(grad):
+            return grad * 2
+
+        class MyMod(torch.nn.Module):
+            def forward(self, x):
+                y = x.mul(2)
+                y.register_hook(my_hook)
+                y.register_hook(my_hook)
+                z = y.mul(3)
+                return (z,)
+
+        mod = MyMod()
+        x0 = torch.ones(4, requires_grad=True)
+        eager_out = mod(x0)
+        eager_out[0].backward(torch.ones(4))
+
+        x1 = torch.ones(4, requires_grad=True)
+        cnts = torch._dynamo.testing.CompileCounterWithBackend("aot_eager")
+        comp_mod = torch._dynamo.optimize(cnts, nopython=True)(mod)
+        comp_out = comp_mod(x1)
+        comp_out[0].backward(torch.ones(4))
+
+        self.assertEqual(cnts.frame_count, 1)
+        my_hook = my_hook2  # noqa: F811
+        self.assertEqual(x0.grad, x1.grad)
+
+        eager_out = mod(x0)
+        eager_out[0].backward(torch.ones(4))
+
+        comp_out = comp_mod(x1)
+
+        self.assertEqual(cnts.frame_count, 2)
+        comp_out[0].backward(torch.ones(4))
+        self.assertEqual(x0.grad, x1.grad)
```

Comment:
Re testing, my recommendation is to grep for register_hook calls in test_autograd.py and adapt those tests into Dynamo tests. (Ideally, we could just Dynamo the original test cases themselves but this may imply yaks.) The idea is that test_autograd.py has all the torture cases for hooks we've accumulated over the years that you wanna check.

Reply:
This is a great suggestion

Reply:
This became a huge yack shave. But what I can do is port some of the more interesting tests to here.

Reply:
that's fine too